### PR TITLE
Upgrade to support PyTorch 2.0+ and Lightning 2.0+

### LINKED ISSUE #376: [Bug] Error raised by version mismatch between PyTorch and Lightning while running tutorial
# 🐛 Bug

PyKale sets PyTorch (>=1.11.0) and Lightning (>=1.3.0,<=1.6.5) in the [setup](https://github.com/pykale/pykale/blob/main/setup.py).
Therefore, in the [digits_dann tutorial](https://colab.research.google.com/github/pykale/pykale/blob/main/examples/digits_dann/tutorial.ipynb), PyKale installs PyTorch 2.0.0 (the latest) and Lightning 1.6.5. However, these two are not compatible, causing Misconfiguration Exception.

## To reproduce
Run [digits_dann tutorial](https://colab.research.google.com/github/pykale/pykale/blob/main/examples/digits_dann/tutorial.ipynb) and `%time trainer.fit(model)` returns the error.

** Stack trace/error message **
```
MisconfigurationException: The provided lr scheduler `LambdaLR` doesn't follow PyTorch's LRScheduler API. You should override the `LightningModule.lr_scheduler_step` hook with your own logic if you are using a custom LR scheduler.
```

## Expected Behaviour
No errors.

## Environment
Default tutorial environments, including
pytorch-lightning-1.6.5
torch-2.0.0+cu118
torchvision-0.15.1+cu118

## Additional context
Two solutions:
1. Fix the PyTorch version <=1.13.1. Simplest but temporal.
2. Update Lightning to 2.x. (may raise issues as discussed in #367)

### PR DESCRIPTION
**Pull request by shuo-zhou**
Fixes #376, #383.

### Description
Upgrade to support PyTorch >= 2.0 and Lightning >= 2.0

1. Change the logging function by moving it into training/validation/test_step. It may affect the outputs of some examples. Please raise an issue when find a negative effect.
2. Change the usage of the [ResNets](https://github.com/pykale/pykale/pull/401/files#diff-a9e038dedae94e5dc004ff253685489d14f94071a7719907fc8f4458765cebdcR153) according to PyTorch 2.0+. 
3. Move the staged optimization of `WDGRLTrainerMod` into [training_step](https://github.com/pykale/pykale/blob/main/kale/pipeline/domain_adapter.py#L872).
4. Add `.readthedocs.yaml` to control the Python version of the docs generation.

Future works: 
1. Remove verbose codes in [ResNets](https://github.com/pykale/pykale/pull/401/files#diff-a9e038dedae94e5dc004ff253685489d14f94071a7719907fc8f4458765cebdcR153).
2. Unify the parameters across the examples, especially how to set the device. (Done in #403 )
3. The accuracy of domain adaptation is currently incorrect. The cause needs to be explored.
4. Remove the commented `optimizer_step` in `WDGRLTrainerMod` after assessing the new staged optimization.

### Status
**Ready**

### Types of changes

- [ ] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] In-line docstrings updated.
- [ ] [Source for documentation at `docs`](https://github.com/pykale/pykale/tree/main/docs/source) manually updated for new API.

### CODE REVIEW

**Comment by xianyuanliu** [comment]
@shuo-zhou Please have a look at commit [16233b6](https://github.com/pykale/pykale/pull/401/commits/16233b6b38a5d8728b962cd58eecb0f9e577e65c). Thanks.

**Comment by shuo-zhou** [comment]
> @shuo-zhou Please have a look at commit [16233b6](https://github.com/pykale/pykale/pull/401/commits/16233b6b38a5d8728b962cd58eecb0f9e577e65c). Thanks.

This commit looks good to me.

**Review by peizhenbai — APPROVED** [review]
LGTM

File: .gitignore [hunk 1]
```diff
@@ -40,6 +40,7 @@ docs/source/backup
 # Testing
 .coverage
 cov*.xml
+test_data
```
Comment (haipinglu) [inline .gitignore:43]: Since we expect test_data to be nested under tests (see below). This was not right in the first place so should not add it here (change the local path instead). Otherwise, this list will be too long.

File: .readthedocs.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  os: ubuntu-22.04
+  tools:
+    python: "3.10"
+
+# Build documentation in the "docs/" directory with Sphinx, Not manually set in the project.
+#sphinx:
+#  configuration: docs/conf.py
+
+# Optional but recommended, declare the Python requirements required
+# to build your documentation
```
Comment (haipinglu) [inline .readthedocs.yaml:18]: newline is not necessary here since we have agreed on line length of 120 (instead of 80).

File: kale/pipeline/domain_adapter.py [hunk 3]
```diff
@@ -47,35 +47,35 @@ def set_requires_grad(model, requires_grad=True):
...
+#             metric_dict[metric_name] = metric_value
+#         else:
+#             metric_dict[metric_name] = metric_value.double().mean()
+#     return metric_dict
+
+
+# # multi-GPUs: mandatory to convert float values into tensors
+# def get_metrics_from_parameter_dict(parameter_dict, device):
+#     """Get a key-value pair from the hyperparameter dictionary"""
+#     return {k: torch.tensor(v, device=device) for k, v in parameter_dict.items()}
```
Comment (haipinglu) [inline kale/pipeline/domain_adapter.py:78]: Is the above no longer needed? Shall we delete rather than comment them off then? Any benefits of keeping them?

File: kale/pipeline/domain_adapter.py [hunk 4]
```diff
@@ -869,66 +801,81 @@ def critic_update_steps(self, batch):
...
+            critic_opt.step()
+            critic_opt.zero_grad()
+
+            # update discriminator opt every k_critic steps
+            if (batch_id + 1) % self._k_critic == 0:
+                task_opt.step()
+                task_opt.zero_grad()
+
+        # log_metrics = get_aggregated_metrics_from_dict(log_metrics)
+        # log_metrics.update(get_metrics_from_parameter_dict(self.get_parameters_watch_list(), loss.device))
```
Comment (haipinglu) [inline kale/pipeline/domain_adapter.py:839]: Again any benefits of keeping rather than removing them?

File: kale/pipeline/domain_adapter.py [hunk 5]
```diff
@@ -869,66 +801,81 @@ def critic_update_steps(self, batch):
...
+    #             optimizer.step()
+    #             optimizer.zero_grad()
+    #
+    #         # update discriminator opt every k_critic steps
+    #         if optimizer_i == 1:
+    #             if (batch_nb + 1) % self._k_critic == 0:
+    #                 optimizer.step()
+    #             optimizer.zero_grad()
+    #
+    #     optimizer.step(closure=second_order_closure)
```
Comment (haipinglu) [inline kale/pipeline/domain_adapter.py:878]: Similar Q: keep vs remove

File: kale/pipeline/domain_adapter.py
Comment (haipinglu) [inline kale/pipeline/domain_adapter.py:1]: Quite heavy changes here. Will this cause any issues with the accuracy?

Also, it will be helpful if such big changes can be explained a big in the PR description so that others can better understand this PR.

File: kale/pipeline/video_domain_adapter.py [hunk 6]
```diff
@@ -324,14 +334,20 @@ def training_step(self, batch, batch_nb):
         else:
             loss = task_loss + self.lamb_da * adv_loss
 
-        log_metrics = get_aggregated_metrics_from_dict(log_metrics)
-        log_metrics.update(get_metrics_from_parameter_dict(self.get_parameters_watch_list(), loss.device))
+        # log_metrics = get_aggregated_metrics_from_dict(log_metrics)
+        # log_metrics.update(get_metrics_from_parameter_dict(self.get_parameters_watch_list(), loss.device))
```
Comment (haipinglu) [inline kale/pipeline/video_domain_adapter.py:338]: The same Q

File: kale/pipeline/video_domain_adapter.py [hunk 7]
```diff
@@ -324,14 +334,20 @@ def training_step(self, batch, batch_nb):
...
         log_metrics["train_total_loss"] = loss
         log_metrics["train_adv_loss"] = adv_loss
         log_metrics["train_task_loss"] = task_loss
 
-        for key in log_metrics:
-            self.log(key, log_metrics[key])
+        # for key in log_metrics:
+        #     self.log(key, log_metrics[key])
```
Comment (haipinglu) [inline kale/pipeline/video_domain_adapter.py:344]: Keep vs remove?

**Review by haipinglu — CHANGES_REQUESTED** [review]
Thanks for this great effort. It will be more efficient to go through the comments in person then make changes. The PR description should summarise the main changes to help others (including future you) keep track.

File: .gitignore (hunk 1 above)
Comment (haipinglu) [reply .gitignore:43]: @shuo-zhou

File: kale/pipeline/domain_adapter.py (hunk 5 above)
Comment (xianyuanliu) [reply kale/pipeline/domain_adapter.py:878]: Keep but add a new comment at the top for description.

`# PyTorch Lightning 2.0+ has a different optimizer_step, so we implement this staged optimization in training_step above. We will retain optimizer_step until we have assessed WDGRLTrainerMod, after which it will be removed.`

File: kale/pipeline/domain_adapter.py
Comment (xianyuanliu) [reply kale/pipeline/domain_adapter.py:1]: The accuracy of domain adaptation is currently incorrect. The cause needs to be explored and may be related to this upgrade. Shuo and I will work on fixing it after merging this PR. 

I'm not 100% sure about other examples. This upgrade mainly changes the logging function by moving it into training/validation/test_step. Theoretically, it should not have any impact.

File: examples/office_multisource_adapt/model.py [hunk 8]
```diff
@@ -68,7 +68,7 @@ def get_model(cfg, dataset, num_channels):
     if cfg.DATASET.NAME.upper() == "DIGITS":
         feature_network = SmallCNNFeature(num_channels)
     else:
-        feature_network = ResNet18Feature(num_channels)
+        feature_network = ResNet18Feature()
```
[the remark this answers is not in the document]
Comment (xianyuanliu) [reply examples/office_multisource_adapt/model.py:71]: ResNet18Feature defaults num_channels to three for images. No need to set it again.

**Review by haipinglu — APPROVED** [review]
Please do a final update of the description before merging. Thanks a lot!
